Stalled RFCs RFC (#130)
NixOS/rfcs: rfcs/0130-stalled-rfcs.md

---
feature: stalled-rfcs
start-date: 2022-06-01
author: kevincox
co-authors: spacekookie

> shepherd-team: (names, to be nominated and accepted by RFC steering committee)

Author: ```suggestion
shepherd-team: @ryantm, @tomberek, @7c6f434c 
```

> shepherd-leader: (name to be appointed by RFC steering committee)

Author: ```suggestion
shepherd-leader: @ryantm

related-issues: (will contain links to implementation PRs)
---

# Summary
[summary]: #summary

This defines two alternate "end states" for RFCs.

The "Awaiting Author" state defines an RFC that isn't ready to be pushed forward and is intentionally stalled. The NixOS RFC Steering Committee will not ping or push this RFC until it returns to the main workflow. These RFCs will be put into a draft status until they are ready to re-enter the process.

The "Lacking Interest" state defines an RFC that can not gather a shepherd team. This is taken to indicate that there is not enough interest in the Nix community. These RFCs will be Closed until more interest is gathered.

# Motivation
[motivation]: #motivation

The NixOS RFC process describes a process with a focus on moving RFCs toward completion. This is where an engaged author and community push an RFC to a decision, either to accept or reject the proposal.

> However things don't always work like this. Sometimes the author doesn't have sufficent time or motivation to keep iterating on the RFC and sometimes the community doesn't have enuogh interested members to staff a shepherd team.

Reviewer: ```suggestion
However things don't always work like this. Sometimes the author doesn't have sufficient time or motivation to keep iterating on the RFC and sometimes the community doesn't have enough interested members to staff a shepherd team.
```

It is important that these "alternate exits" are well defined to avoid wasting effort trying to push a stalled RFC forward and to set clear expectations for all participants.

TL;DR timeouts are critical in distributed systems.

# Detailed design
[design]: #detailed-design

Here is an updated diagram for the RFC process.

> ```mermaid

Reviewer 2: The rendered graph looks incorrect; "On Hold" also goes through "Author Withdraws" and ends at "Closed - Withdrawn", but both of those sound wrong to me. Shouldn't there be a separate end state for "On Hold"?

Author: Hmm, you are right but I don't know how to fix this. Suggestions welcome.

Author: cc @ryantm Let's keep the diagram discussion here.

I'm not sure if mermaid supports manual positioning so I see a few options:

1. Ignore the rendering bug.
2. Change `mermaid` to `text` so that it doesn't render and the source is the visible thing.
3. Recreate the diagram with another tool. (Harder to edit unfortunately).
4. Drop the diagram completely.
5. Block this on a mermaid rendering bug.

Personally I am in favour of 1 but if people feel strongly I could do 2 or 3.

Reviewer 3: It's just awkward rendering -- the "On Hold" label corresponds to the arrow to its left (from "Discuss and Refine" to "Draft"). See my suggestion for a hacky fix.

Author: Thanks, that's a clever workaround for lack of manual positioning 🙂

graph TD
    Start((Start)) --> Draft
    Draft -->|Ready for Review| Discuss
    style Draft fill:#008,color:#FFF

    Discuss[Discuss and Refine]

>     Discuss --> |On Hold| Draft

Reviewer 3: ```suggestion
    Discuss ---> |On Hold| Draft
```
This results in the label being closer to the arrow.

    Discuss --> |Can't Find Shepherds| NoShepherds
    Discuss --> |Motion for FCP| FCP

    NoShepherds[Closed - Lack of Interest]:::closed
    NoShepherds --> |Renewed Interest| Discuss

    FCP[Final Coment Phase]
    FCP --> |FCP Canceled| Discuss
    FCP --> |Accept| Merged
    FCP --> |Reject| Rejected
    
    Merged
    style Merged fill:#080,color:#FFF
    
    Rejected[Closed - Rejected]:::closed

    Withdrawn[Closed - Withdrawn]:::closed
    Discuss & Draft --->|Author Withdraws| Withdrawn

    classDef closed fill:#800,color:#FFF
```

There are 3 new transitions in this diagram.

## On Hold

> This is for cases where the author is unable or unwilling to update the RFC in a timely fashon and would like to indicate that the RFC should not continue to be pushed forward for the time being.

Reviewer: ```suggestion
This is for cases where the author is unable or unwilling to update the RFC in a timely fashion and would like to indicate that the RFC should not continue to be pushed forward for the time being.
```

The author can [mark the PR as a Draft](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request#converting-a-pull-request-to-a-draft) and the existing label [status: on hold](https://github.com/NixOS/rfcs/pulls?q=is%3Apr+label%3A%22status%3A+on+hold%22+is%3Aclosed) can be applied to the PR by the NixOS RFC Steering Committee for tracking reasons.

RFCs in this status will be completely ignored by the NixOS RFC Steering Committee. It is also expected that community members will largely refrain from review until the author has indicated that the RFC is again ready for review.

## Can't Find Shepherds

> If an RFC is unable to find adaqite number of shepherds in 2 months it will be considered "low interest". The RFC will be closed to indicate that there is no expection of continuous forward progress.

Reviewer: ```suggestion
If an RFC is unable to find adequate number of shepherds in 2 months it will be considered "low interest". The RFC will be closed to indicate that there is no expectation of continuous forward progress.
```

> After 1 month of being open the following message will be added to the PR to suggest ways to find more support:

Reviewer: ```suggestion
After 1 month of being open, the following message will be added to the PR to suggest ways to find more support:
```

> > This RFC has not acquired enough shepherds. This typically shows lack of interest from the community. In order to progress [a full shepherd team is required](https://github.com/NixOS/rfcs/#shepherd-team). Consider trying to raise interest by [posting in Discourse](https://discourse.nixos.org/), [taking in Matrix](https://matrix.to/#/#community:nixos.org) or reaching out to people that you know.

Reviewer: ```suggestion
> This RFC has not acquired enough shepherds. This typically shows lack of interest from the community. In order to progress [a full shepherd team is required](https://github.com/NixOS/rfcs/#shepherd-team). Consider trying to raise interest by [posting in Discourse](https://discourse.nixos.org/), [talking in Matrix](https://matrix.to/#/#community:nixos.org) or reaching out to people that you know.
```

>
> If not enough shepherds can be found in the next month we will close this RFC until we can find enough interested participants. The PR can be reopened at any time if more shepherd nominations are made.
>
> [See more info on the Nix RFC process here](https://github.com/NixOS/rfcs/blob/master/README.md)

> If another month has elapsed since the warning without sufficent accepted nominations the following message will be posted, the `status: insufficent interest` label will be added and the issue will be closed.

Reviewer: ```suggestion
If another month has elapsed since the warning without sufficient accepted nominations the following message will be posted, the `status: insufficient interest` label will be added and the issue will be closed.
```

> This RFC is being closed due to lack interest. If enough shepherds are found this issue can be reopened. If you don't have permission to reopen please [open an issue for the NixOS RFC Steering Committee](https://github.com/NixOS/rfc-steering-committee/issues/new) linking to this PR.

## Renewed Interest

> If an RFC that was closed due to lack of interest finds sufficent interest it can be reopened and the `status: insufficent interest` label removed. The RFC will then continue through the RFC process as normal.

Reviewer: ```suggestion
If an RFC that was closed due to lack of interest finds sufficient interest it can be reopened and the `status: insufficient interest` label removed. The RFC will then continue through the RFC process as normal.
```

# Examples and Interactions
[examples-and-interactions]: #examples-and-interactions

## More Research Required

> If an RFC is opened and discussion reveals that more research is required the author may now return their RFC to draft status to indicate that they are working on an update.

Reviewer: ```suggestion
If an RFC is opened and discussion reveals that more research is required, the author may now return their RFC to draft status to indicate that they are working on an update.
```

> This is already done frequently done today but this RFC codifies this process. The alternative which is sometimes seen is that the NixOS RFC Steering Committee keeps pinging the RFC for months trying to move the RFC forward which is annoying and unproductive.

Reviewer: ```suggestion
This is already frequently done today, but this RFC codifies this process. The alternative, which is sometimes seen is that the NixOS RFC Steering Committee keeps pinging the RFC for months trying to move the RFC forward which is annoying and unproductive.
```

## Can't Find Shepherds

> If an RFC is opened but a team of shepherds can't be assembed the RFC will now be closed.

Reviewer: ```suggestion
If an RFC is opened but a team of shepherds can't be assembled the RFC will now be closed.
```

Currently there is no process here and the NixOS RFC Steering Committee is unsure what to do and tends to annoyingly ping these RFCs with no positive outcome for a few months then improvise. This RFC proposes a well-defined process to ensure that everyone understands the process.

# Drawbacks
[drawbacks]: #drawbacks

## Closing a PR may be demotivating.

> Strugging to find shepherds for an RFC may already be demotivating and having the RFC closed can be an aditional flag that the RFC is strugging. This may be demoralizing to the RFC author.

Reviewer: ```suggestion
Struggling to find shepherds for an RFC may already be demotivating and having the RFC closed can be an additional flag that the RFC is struggling. This may be demoralizing to the RFC author.
```

> This issue is mitigated by having clear and well-thought-out messages to make it clear what is occuring at this stage of the process. This is hopefully better overall than the previously poorly defined state of forever waiting for more shepherds to appear.

Reviewer: ```suggestion
This issue is mitigated by having clear and well-thought-out messages to make it clear what is occurring at this stage of the process. This is hopefully better overall than the previously poorly defined state of forever waiting for more shepherds to appear.
```

## Closing a PR may reduce visibility for the RFC.

Leaving PRs open may increase the chance that a potential shepherd finds the RFC. Closing RFCs that can't find shepherds may worsen the thing that ails them.

> It is the author's opinion that having dozens of open RFCs will also reduce visibility from the ones that do have more interest and are more likely to find a full shepherd team. Therefore this is considered acceptable cost. The addition of the `status: insufficent interest` label may also help to bring attention to these RFCs that could benifit from would-be shepherds.

Reviewer: ```suggestion
It is the author's opinion that having dozens of open RFCs will also reduce visibility from the ones that do have more interest and are more likely to find a full shepherd team. Therefore this is considered an acceptable cost. The addition of the `status: insufficient interest` label may also help to bring attention to these RFCs that could benefit from would-be shepherds.
```

# Alternatives
[alternatives]: #alternatives

## Do Nothing

Doing nothing leaves the behaviour to be taken when an RFC can not find a shepherd team undefined. This leads to inconsistent handling by the NixOS RFC Steering Committee and leaves community members unable to understand what to expect.

## Leave the PR open and add a label

This can reduce the burden from the NixOS RFC Steering Committee however it still clutters the list of "Open" RFCs with these RFCs that are not seeing forward progress. Ultimately it is a decision of definition of what "Open" means and this RFC takes the stance that if an RFC is stalled for too long it makes sense to remove it from the default search for "Open" RFCs.

> 

Author: ```suggestion

Due to similarity with [RFC 51 - Mark stale nixpkgs issues)[https://github.com/NixOS/rfcs/pull/51] which had a lot of backlash, extra effort was made to ensure that this was not too controversial. Notably a form post at https://discourse.nixos.org/t/stalled-rfcs-proposing-rfcs-with-no-interest/21952. It appears that to the Nix community this is in fact a substantially different case so closing is acceptable. 

```

Reviewer: "forum post" typo. I think the main reason this is different from 51 is that it will be people exercising their judgement while manually closing them.

Reviewer 4: I would probably stress not the backlash but «settled on labelling without closing after a lot of discussion» to point to the relevant part.

Otherwise I agree.

Author: Thanks for the suggestions. I've incorporated them all. Please double check.

> This can reduce the burden from the NixOS RFC Steering Committee however it still clutters the list of "Open" RFCs with these RFCs that are not seeing forward progress. Ultimately it is a decision of definition of what "Open" means and this RFC takes the stance that if an RFC is stalled for too long it makes sense to remove it from the default search for "Open" RFCs.

Reviewer 4: Argument in favour of this approach: we have had _two_ separate discussions about stale PR handling, with the end result being «label but don't close». I think consistency is good.

Note that we have few enough RFCs that a round of RFC SC pings ensures that probably-live ones are indeed on top in the list sorted by last-updated.

Author: The main tangible difference I see is if we want these to appear in the default search which on GitHub is "open order by created descending". The current proposal will hide these whereas this alternative will interleave these with other open and active PRs.

RFC SC can easily pin the right search to see what they want so it doesn't really matter to them.

Personally I have a small preference for hiding them to focus attention on "active" PRs but I definitely see the argument for trying to get more attention to these PRs that need attention most, even if that thins out the attention for other PRs a bit.

But I don't have a strong opinion either way. I avoided responding to this comment earlier in hopes that others would add their opinion but it seems like there isn't that much debate here.

Reviewer 4: We have RFCs#51 where various «clean list» argument were presentde but ended up rejected, and RFCs#124 seems to show that sometimes there is a lot of support to make policy similar to RFCs#51. It _is_ true that RFC proposals are slighlty different, but then I guess we need a detailed argument why it is better to pick a different policy than for Nixpkgs and Nix repos here, and I guess a dedicated large announcement. Otherwise it would be natural to expect that things get labeled not closed.

I would expect people look for «what is active» and sort by activity, or «what discussions have I missed» (sorting by creation time) and then it is not always clear if RFCs lacking shepherds are out of scope.

It does look like right now «draft» label seems to mix waiting-on-author and lacking-shepherds right now… making this available at glance will surely be useful.

For closing clearly-not-going-anywhere… maybe add a transition where force closing can be done if some _person_ bothers to ask the author what's the plan and does not get any response at all in a month? I guess switch of authorship requires resubmission in any case.

Author: I'm usually against "clean list" arguments. But that is more because for issues it is helpful to find the existing ones. For RFCs I'm not as sure, but maybe I'm over-valuing the "browsing" use case.

I think if we do the label approach then we end up where the author may give up after a while and close. Or we may have lots of open PRs where the author gave up and forgot.

Do you feel strongly?

Reviewer 4: > for issues

I would say RFCs are also proposed as a reaction to a perceived issue with something we are doing!

> maybe I'm over-valuing the "browsing" use case.

I am making an even weaker objection: it is unclear how this use case is split among a few plausible variants (which are better served by different decisions).

> I think if we do the label approach then we end up where the author may give up after a while and close. Or we may have lots of open PRs where the author gave up and forgot.

Author withdrawing is fine. As I said, human-judgement-driven (without recommended timeline) check if author went missing is also fine and maybe it could be added as a text-only clarification (if authors are nowhere to be found, I guess it counts as the authors withdrawing themselves from the process, so it's just a footnote for an existing transition…).

> Do you feel strongly?

I feel strongly that fixed-delay-based closing of discussions within NixOS namespace at GitHub needs a very public announcement of intent and justification which explicitly shows that arguments from #51 are not applicable here (and then, well, lack of strong pushback in replies to that announcement).

But my position is more process-based: I have nothing against this decision if the heads-up announcement is there, stresses the difference, and is well received. (I guess I would prefer uniformity, but that's a weak preference and people mostly agreeing this case is separate would override it)

Author: I'd be happy to get more eyes on this but am not sure of the best avenue to get more eyes and opinions than this RFC What would you recommend?

Reviewer 4: Maybe a Discource heads-up announcement that one part of the current RFC might be surprising for some? If you think it is a good idea, we could mention-ping the other shepherds to check for reasonability, maybe in the non-line-tied part of the discussion.

Author: Attempted to solicit more feedback here: https://discourse.nixos.org/t/stalled-rfcs-proposing-rfcs-with-no-interest/21952

Reviewer 3: (reposted from where I previously put it, which was the wrong place)
I think an important distinction between auto-closing issues and this where the responsibility for getting it out of that state lies; with automatically closed issues, only people with write access to the repo can reopen them. In the case of RFCs, any group of people with enough interest can pick the RFC up and get it moving again, and the RFC Steering Committee will see and apply the administrivial changes.

Author: @7c6f434c and others. I have proposed a note to codify this check in the RFC document. Does this sound good to you?

https://github.com/NixOS/rfcs/pull/130/files#r990635602
